src/supernova/mod.rs

```diff
@@ -0,0 +1,1194 @@
+  /// Returns the number of variables in the primary and secondary circuits
+  pub fn num_variables(&self) -> (usize, usize) {
+    (
+      self.r1cs_shape_primary.num_vars,
+      self.r1cs_shape_secondary.num_vars,
+    )
+  }
+}
+
+/*
+ SuperNova takes Ui a list of running instances.
+ One instance of Ui is a struct called RunningClaim.
+*/
+pub struct RunningClaim<G1, G2, Ca, Cb>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  Ca: StepCircuit<G1::Scalar>,
+  Cb: StepCircuit<G2::Scalar>,
+{
+  _phantom: PhantomData<G1>,
+  claim: Ca,
+  circuit_secondary: Cb,
+  params: PublicParams<G1, G2>,
+}
+
+impl<G1, G2, Ca, Cb> RunningClaim<G1, G2, Ca, Cb>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  Ca: StepCircuit<G1::Scalar>,
+  Cb: StepCircuit<G2::Scalar>,
+{
+  pub fn new(circuit_primary: Ca, circuit_secondary: Cb, num_augmented_circuits: usize) -> Self {
+    let claim = circuit_primary.clone();
+
+    let pp = PublicParams::<G1, G2>::setup_without_commitkey(
+      claim.clone(),
+      circuit_secondary.clone(),
+      num_augmented_circuits,
+    );
+
+    Self {
+      _phantom: PhantomData,
+      claim,
+      circuit_secondary: circuit_secondary.clone(),
+      params: pp,
+    }
+  }
+}
+
+/// A SNARK that proves the correct execution of an non-uniform incremental computation
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct RecursiveSNARK<G1, G2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+{
+  r_W_primary: Vec<Option<RelaxedR1CSWitness<G1>>>,
+  r_U_primary: Vec<Option<RelaxedR1CSInstance<G1>>>,
+  r_W_secondary: Option<RelaxedR1CSWitness<G2>>,
+  r_U_secondary: Option<RelaxedR1CSInstance<G2>>,
+  l_w_secondary: R1CSWitness<G2>,
+  l_u_secondary: R1CSInstance<G2>,
+  pp_digest: G1::Scalar,
+  i: usize,
+  zi_primary: Vec<G1::Scalar>,
+  zi_secondary: Vec<G2::Scalar>,
+  program_counter: G1::Scalar,
+  augmented_circuit_index: usize,
+}
+
+impl<G1, G2> RecursiveSNARK<G1, G2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+{
+  /// Create a new `RecursiveSNARK` (or updates the provided `RecursiveSNARK`)
+  /// by executing a step of the incremental computation
+  pub fn prove_step<C1: StepCircuit<G1::Scalar>, C2: StepCircuit<G2::Scalar>>(
+    circuit_index: usize,
+    num_augmented_circuits: usize,
+    program_counter: G1::Scalar,
+    pp: &PublicParams<G1, G2>,
+    recursive_snark: Option<Self>,
+    c_primary: C1,
+    c_secondary: C2,
+    z0_primary: &Vec<G1::Scalar>,
+    z0_secondary: &Vec<G2::Scalar>,
+    ck_primary: &CommitmentKey<G1>,
+    ck_secondary: &CommitmentKey<G2>,
+  ) -> Result<Self, NovaError> {
+    if z0_primary.len() != pp.F_arity_primary || z0_secondary.len() != pp.F_arity_secondary {
+      return Err(NovaError::InvalidInitialInputLength);
+    }
+
+    match recursive_snark {
```

Comment:
This code might be based on an older version of src/lib.rs? I think there was some non-trivial improvement from using mutable types. Is it worth porting that code here?

Reply:
FYI, this is about #163

Reply:
Thanks for the hints. Fixed in latest commit and align with Nova performance fix PR https://github.com/microsoft/Nova/pull/163